certs/certs.go

```diff
@@ -0,0 +1,293 @@
+	// the power table used in the next instance.
+	SupplementalData gpbft.SupplementalData
+	// Indexes in the base power table of the certifiers (bitset)
+	Signers bitfield.BitField
+	// Aggregated signature of the certifiers
+	Signature []byte
+	// Changes between the power table used to validate this finality certificate and the power
+	// used to validate the next finality certificate. Sorted by ParticipantID, ascending.
+	PowerTableDelta []PowerTableDelta
+}
+
+// NewFinalityCertificate constructs a new finality certificate from the given power delta (from
+// `MakePowerTableDiff`) and justification (from GPBFT).
+//
+// Note, however, that this function does not attempt to validate the resulting finality
+// certificate (beyond verifying that it is a justification for the correct round). You can do so by
+// immediately calling `ValidateFinalityCertificates` on the result.
+func NewFinalityCertificate(powerDelta []PowerTableDelta, justification *gpbft.Justification) (*FinalityCertificate, error) {
+	if justification.Vote.Step != gpbft.DECIDE_PHASE {
+		return nil, xerrors.Errorf("can only create a finality certificate from a decide vote, got phase %s", justification.Vote.Step)
+	}
+
+	if justification.Vote.Round != 0 {
+		return nil, xerrors.Errorf("expected decide round to be 0, got round %d", justification.Vote.Round)
+	}
+
+	if justification.Vote.Value.IsZero() {
+		return nil, xerrors.Errorf("got a decision for bottom for instance %d", justification.Vote.Instance)
+	}
+
+	return &FinalityCertificate{
+		GPBFTInstance:    justification.Vote.Instance,
+		SupplementalData: justification.Vote.SupplementalData,
+		ECChain:          justification.Vote.Value,
+		Signers:          justification.Signers,
+		Signature:        justification.Signature,
+		PowerTableDelta:  powerDelta,
+	}, nil
+}
+
+// ValidateFinalityCertificates validates zero or more finality certificates, returning the next
+// instance after the last valid finality certificates, any newly finalized tipsets, and the next
+// power table to use as-of the last valid finality certificate. If passed a non-nil `base` tipset,
+// validate that the finalized chain starts with that tipset (accept any finalized chain otherwise).
+//
+// Returns an error if it encounters any invalid finality certificates, along with the last valid
+// instance, finalized chain epochs, etc. E.g., if provided a partially valid chain of finality
+// certificates, this function will return a (possibly empty) prefix of the EC chain correctly
+// finalized, the instance of the first invalid finality certificate, and the power table that
+// should be used to validate that finality certificate, along with the error encountered.
+func ValidateFinalityCertificates(verifier gpbft.Verifier, network gpbft.NetworkName, prevPowerTable gpbft.PowerEntries, nextInstance uint64, base *gpbft.TipSet,
+	certs ...FinalityCertificate) (_nextInstance uint64, chain gpbft.ECChain, newPowerTable gpbft.PowerEntries, err error) {
```

Comment:
Channelling @masih , consider introducing structs for these high arity parameter and return types.